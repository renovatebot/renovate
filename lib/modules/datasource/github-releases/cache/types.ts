/**
 * Every `AbstractGithubDatasourceCache` implementation
 * should have `graphqlQuery` that uses parameters
 * defined this interface.
 */
export interface GithubQueryParams {
  owner: string;
  name: string;
  cursor: string | null;
  count: number;
}

/**
 * Every `AbstractGithubDatasourceCache` implementation
 * should have `graphqlQuery` that resembles the structure
 * of this interface.
 */
export interface QueryResponse<T = unknown> {
  repository: {
    payload: {
      nodes: T[];
      pageInfo: {
        hasNextPage: boolean;
        endCursor: string;
      };
    };
  };
}

/**
 * Base interface meant to be extended by all implementations.
 * Must have `version` and `releaseTimestamp` fields.
 */
export interface StoredItemBase {
  /** The values of `version` field meant to be unique. */
  version: string;

  /** The `releaseTimestamp` field meant to be ISO-encoded date. */
  releaseTimestamp: string;
}

/**
 * The data structure stored in the package cache.
 */
export interface GithubDatasourceCache<StoredItem extends StoredItemBase> {
  items: Record<string, StoredItem>;

  /** Cache full reset decision is based on `createdAt` value. */
  createdAt: string;

  /** Cache soft updates are performed depending on `updatedAt` value. */
  updatedAt: string;

  /** Latest release timestamp (`releaseTimestamp`) of all releases. */
  lastReleasedAt?: string;
}

/**
 * The configuration for cache.
 */
export interface CacheOptions {
  /**
   * How many minutes to wait until next cache update
   */
  updateAfterMinutes?: number;

  /**
   * If package was released recently, we assume higher
   * probability of having one more release soon.
   *
   * In this case, we use `updateAfterMinutesFresh` option.
   */
  packageFreshDays?: number;

  /**
   * If package was released recently, we assume higher
   * probability of having one more release soon.
   *
   * In this case, this option will be used
   * instead of `updateAfterMinutes`.
   *
   * Fresh period is configured via `freshDays` option.
   */
  updateAfterMinutesFresh?: number;

  /**
   * How many days to wait until full cache reset (for single package).
   */
  resetAfterDays?: number;

  /**
   * Delays cache reset by some random amount of minutes,
   * in order to stabilize load during mass cache reset.
   */
  resetDeltaMinutes?: number;

  /**
   * How many days ago the package should be published to be considered as stable.
   * Since this period is expired, it won't be refreshed via soft updates anymore.
   */
  unstableDays?: number;

  /**
   * How many items per page to obtain per page during initial fetch (i.e. pre-fetch)
   */
  itemsPerPrefetchPage?: number;

  /**
   * How many pages to fetch (at most) during the initial fetch (i.e. pre-fetch)
   */
  maxPrefetchPages?: number;

  /**
   * How many items per page to obtain per page during the soft update
   */
  itemsPerUpdatePage?: number;

  /**
   * How many pages to fetch (at most) during the soft update
   */
  maxUpdatePages?: number;
}
