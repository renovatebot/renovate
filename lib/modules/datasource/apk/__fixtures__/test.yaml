# Test fixture for APK datasource
# This represents an Alpine Linux package repository structure

# Example repository URLs that would be used:
# - https://dl-cdn.alpinelinux.org/alpine/v3.19/main
# - https://dl-cdn.alpinelinux.org/alpine/v3.19/community
# - https://packages.wolfi.dev/os

# Example package entries that would be found in APKINDEX.tar.gz:
# P:nginx
# V:1.24.0-r0
# T:nginx web server
# U:https://nginx.org/
# S:1234567
# I:1234567890
# o:nginx
# A:x86_64
# L:BSD-2-Clause
# D:openssl
# p:nginx
# c:
# r:

# Example packages that would be extracted:
packages:
  - name: nginx
    version: 1.24.0-r0
    description: nginx web server
    url: https://nginx.org/
    size: 1234567
    buildDate: 1234567890
    origin: nginx
    arch: x86_64
    license: BSD-2-Clause
    depends: [openssl]
    provides: [nginx]
    conflicts: []
    replaces: []
