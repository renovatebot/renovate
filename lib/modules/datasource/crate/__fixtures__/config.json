{
  "dl": "https://crates.io/api/v1/crates",
  "api": "https://crates.io"
}
