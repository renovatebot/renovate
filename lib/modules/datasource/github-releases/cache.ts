import { DateTime, DurationLike } from 'luxon';
import { logger } from '../../../logger';
import * as packageCache from '../../../util/cache/package';
import type {
  GithubGraphqlResponse,
  GithubHttp,
} from '../../../util/http/github';
import type { GetReleasesConfig, Release } from '../types';
import { getApiBaseUrl } from './common';

const softReset: DurationLike = { minutes: 30 };
const hardReset: DurationLike = { days: 7 };
const stabilityPeriod: DurationLike = { months: 1 };

const query = `
query(
  $owner: String!,
  $name: String!,
  $cursor: String,
  $count: Int!
) {
  repository(owner: $owner, name: $name) {
    releases(after: $cursor, first: $count, orderBy: {field: CREATED_AT, direction: DESC}) {
      nodes {
        version: tagName
        gitRef: tagName
        releaseTimestamp: publishedAt
        isDraft
        isPrerelease
        updatedAt
      }

      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
}
`;

interface QueryParams {
  owner: string;
  name: string;
  cursor: string | null;
  count: number;
}

interface QueryResponse {
  repository: {
    releases: {
      nodes: FetchedItem[];
      pageInfo: {
        hasNextPage: boolean;
        endCursor: string;
      };
    };
  };
}

interface ItemBase {
  version: string;
  releaseTimestamp: string;
}

interface FetchedItem extends ItemBase {
  isDraft: boolean;
  isPrerelease: boolean;
  updatedAt: string;
}

interface CachedItem extends ItemBase {
  gitRef: string;
  isStable?: boolean;
}

interface GithubReleasesCache {
  releases: Record<string, CachedItem>;
  releasesUpdatedAt: Record<string, string>;
  createdAt: string;
  updatedAt: string;
}

function coerceFetchedItem(item: FetchedItem): CachedItem {
  const { version, releaseTimestamp, isDraft, isPrerelease } = item;
  const gitRef = version;
  const result: CachedItem = { version, gitRef, releaseTimestamp };
  if (isPrerelease || isDraft) {
    result.isStable = false;
  }
  return result;
}

function createCache(now: DateTime): GithubReleasesCache {
  return {
    releases: {},
    releasesUpdatedAt: {},
    createdAt: now.toISO(),
    updatedAt: now.minus(softReset).toISO(),
  };
}

function isExpired(
  now: DateTime,
  date: string,
  duration: DurationLike
): boolean {
  const then = DateTime.fromISO(date);
  return now >= then.plus(duration);
}

export async function getCacheableReleases(
  http: GithubHttp,
  config: GetReleasesConfig
): Promise<Release[]> {
  const now = DateTime.now();
  let cache: GithubReleasesCache = createCache(now);
  const { packageName, registryUrl } = config;

  const baseUrl = getApiBaseUrl(registryUrl).replace('/v3/', '/');
  const [owner, name] = packageName.split('/');
  if (owner && name) {
    const cacheNs = `datasource-github-releases-graphql`;
    const cacheKey = `${baseUrl}:${owner}/${name}`;
    const cachedRes = await packageCache.get<GithubReleasesCache>(
      cacheNs,
      cacheKey
    );

    let isCacheUpdated = false;

    if (cachedRes && !isExpired(now, cachedRes.createdAt, hardReset)) {
      cache = cachedRes;
    } else {
      isCacheUpdated = true;
    }

    if (isExpired(now, cache.updatedAt, softReset)) {
      const variables: QueryParams = {
        owner,
        name,
        cursor: null,
        count: 100,
      };

      const checkedItems = new Set<string>();

      try {
        let pagesAllowed = 10;
        let isIterating = true;
        while (pagesAllowed > 0 && isIterating) {
          const graphqlRes = await http.postJson<
            GithubGraphqlResponse<QueryResponse>
          >('/graphql', {
            baseUrl,
            body: { query, variables },
          });
          pagesAllowed -= 1;

          const data = graphqlRes.body.data;
          if (data) {
            const {
              nodes: fetchedItems,
              pageInfo: { hasNextPage, endCursor },
            } = data.repository.releases;

            if (!hasNextPage) {
              isIterating = false;
            } else {
              variables.cursor = endCursor;
            }

            for (const item of fetchedItems) {
              const { version, releaseTimestamp, updatedAt } = item;
              checkedItems.add(version);

              const storedItem = cache.releases[version];
              const storedItemUpdatedAt = cache.releasesUpdatedAt[version];
              if (!storedItem || storedItemUpdatedAt !== updatedAt) {
                cache.releases[version] = coerceFetchedItem(item);
                isCacheUpdated = true;
              } else if (isExpired(now, releaseTimestamp, stabilityPeriod)) {
                isIterating = false;
                break;
              }
            }
          }
        }
      } catch (err) {
        logger.debug(
          { err },
          `GitHub releases: error fetching cacheable GraphQL data`
        );
      }

      for (const [version, item] of Object.entries(cache.releases)) {
        if (
          !isExpired(now, item.releaseTimestamp, stabilityPeriod) &&
          !checkedItems.has(version)
        ) {
          delete cache.releases[version];
          isCacheUpdated = true;
        }
      }

      if (isCacheUpdated) {
        const expiry = DateTime.fromISO(cache.createdAt).plus(hardReset);
        const { minutes: ttlMinutes } = expiry
          .diff(now, ['minutes'])
          .toObject();
        if (ttlMinutes && ttlMinutes > 0) {
          cache.updatedAt = now.toISO();
          await packageCache.set(cacheNs, cacheKey, cache, ttlMinutes);
        }
      }
    }
  }

  const result = Object.values(cache.releases);
  return result;
}
