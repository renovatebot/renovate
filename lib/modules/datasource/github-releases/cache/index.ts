import type {
  CacheOptions,
  GithubCachedRelease,
  GithubGraphqlRelease,
} from '../../../../util/github/types';
import type { GithubHttp } from '../../../../util/http/github';
import { AbstractGithubDatasourceCache } from './cache-base';

export const query = `
query ($owner: String!, $name: String!, $cursor: String, $count: Int!) {
  repository(owner: $owner, name: $name) {
    payload: releases(
      first: $count
      after: $cursor
      orderBy: {field: CREATED_AT, direction: DESC}
    ) {
      nodes {
        version: tagName
        releaseTimestamp: publishedAt
        isDraft
        isPrerelease
        url
        id: databaseId
        name
        description
      }
      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
}
`;

export class CacheableGithubReleases extends AbstractGithubDatasourceCache<
  GithubCachedRelease,
  GithubGraphqlRelease
> {
  cacheNs = 'github-datasource-graphql-releases';
  graphqlQuery = query;

  constructor(http: GithubHttp, opts: CacheOptions = {}) {
    super(http, opts);
  }

  coerceFetched(item: GithubGraphqlRelease): GithubCachedRelease | null {
    const {
      version,
      releaseTimestamp,
      isDraft,
      isPrerelease,
      url,
      id,
      name,
      description,
    } = item;

    if (isDraft) {
      return null;
    }

    const result: GithubCachedRelease = {
      version,
      releaseTimestamp,
      url,
      id,
      name,
      description,
    };

    if (isPrerelease) {
      result.isStable = false;
    }

    return result;
  }
}
