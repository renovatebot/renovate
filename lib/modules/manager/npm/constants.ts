export const NPM_CACHE_DIR = 'npm';
export const PNPM_CACHE_BASE_DIR = 'pnpm';
export const PNPM_STORE_DIR = `${PNPM_CACHE_BASE_DIR}/store`;
export const PNPM_CACHE_DIR = `${PNPM_CACHE_BASE_DIR}/cache`;
export const YARN_CACHE_DIR = 'yarn';
export const YARN_GLOBAL_DIR = 'berry';
