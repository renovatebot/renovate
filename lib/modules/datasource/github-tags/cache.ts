import type {
  CacheOptions,
  GithubCachedTag,
  GithubGraphqlTag,
} from '../../../util/github/types';
import type { GithubHttp } from '../../../util/http/github';
import { AbstractGithubDatasourceCache } from '../github-releases/cache/cache-base';

const query = `
query ($owner: String!, $name: String!, $cursor: String, $count: Int!) {
  repository(owner: $owner, name: $name) {
    payload: refs(
      first: $count
      after: $cursor
      orderBy: {field: TAG_COMMIT_DATE, direction: DESC}
      refPrefix: "refs/tags/"
    ) {
      nodes {
        version: name
        target {
          type: __typename
          ... on Commit {
            hash: oid
            releaseTimestamp: committedDate
          }
          ... on Tag {
            target {
              ... on Commit {
                hash: oid
              }
            }
            tagger {
              releaseTimestamp: date
            }
          }
        }
      }
      pageInfo {
        hasNextPage
        endCursor
      }
    }
  }
}
`;

export class CacheableGithubTags extends AbstractGithubDatasourceCache<
  GithubCachedTag,
  GithubGraphqlTag
> {
  readonly cacheNs = 'github-datasource-graphql-tags-v2';
  readonly graphqlQuery = query;

  constructor(http: GithubHttp, opts: CacheOptions = {}) {
    super(http, opts);
  }

  coerceFetched(item: GithubGraphqlTag): GithubCachedTag | null {
    const { version, target } = item;
    if (target.type === 'Commit') {
      const { hash, releaseTimestamp } = target;
      return { version, hash, releaseTimestamp };
    } else if (target.type === 'Tag') {
      const { hash } = target.target;
      const { releaseTimestamp } = target.tagger;
      return { version, hash, releaseTimestamp };
    }
    return null;
  }
}
