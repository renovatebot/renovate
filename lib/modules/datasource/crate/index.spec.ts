import delay from 'delay';
import fs from 'fs-extra';
import _simpleGit, { SimpleGit } from 'simple-git';
import { DirectoryResult, dir } from 'tmp-promise';
import { dirname, join } from 'upath';
import { getPkgReleases } from '..';
import { Fixtures } from '../../../../test/fixtures';
import * as httpMock from '../../../../test/http-mock';
import { GlobalConfig } from '../../../config/global';
import type { RepoGlobalConfig } from '../../../config/types';
import { EXTERNAL_HOST_ERROR } from '../../../constants/error-messages';
import * as memCache from '../../../util/cache/memory';
import type { RegistryInfo } from './types';
import { CrateDatasource } from '.';

jest.mock('simple-git');
const simpleGit: jest.Mock<Partial<SimpleGit>> = _simpleGit as never;

const API_BASE_URL = CrateDatasource.CRATES_IO_API_BASE_URL;

const baseUrl =
  'https://raw.githubusercontent.com/rust-lang/crates.io-index/master/';

const datasource = CrateDatasource.id;

function setupGitMocks(delayMs?: number): { mockClone: jest.Mock<any, any> } {
  const mockClone = jest
    .fn()
    .mockName('clone')
    .mockImplementation(
      async (_registryUrl: string, clonePath: string, _opts) => {
        if (delayMs && delayMs > 0) {
          await delay(delayMs);
        }

        const path = `${clonePath}/my/pk/mypkg`;
        fs.mkdirSync(dirname(path), { recursive: true });
        fs.writeFileSync(path, Fixtures.get('mypkg'), { encoding: 'utf8' });
      }
    );

  simpleGit.mockReturnValue({
    clone: mockClone,
  });

  return { mockClone };
}

function setupErrorGitMock(): { mockClone: jest.Mock<any, any> } {
  const mockClone = jest
    .fn()
    .mockName('clone')
    .mockImplementation((_registryUrl: string, _clonePath: string, _opts) =>
      Promise.reject(new Error('mocked error'))
    );

  simpleGit.mockReturnValue({
    clone: mockClone,
  });

  return { mockClone };
}

function mockCratesApiCallFor(crateName: string, response?: httpMock.Body) {
  httpMock
    .scope(API_BASE_URL)
    .get(`/crates/${crateName}?include=`)
    .reply(response ? 200 : 404, response);
}

describe('modules/datasource/crate/index', () => {
  describe('getIndexSuffix', () => {
    it('returns correct suffixes', () => {
      expect(CrateDatasource.getIndexSuffix('a')).toStrictEqual(['1', 'a']);
      expect(CrateDatasource.getIndexSuffix('1')).toStrictEqual(['1', '1']);
      expect(CrateDatasource.getIndexSuffix('1234567')).toStrictEqual([
        '12',
        '34',
        '1234567',
      ]);
      expect(CrateDatasource.getIndexSuffix('ab')).toStrictEqual(['2', 'ab']);
      expect(CrateDatasource.getIndexSuffix('abc')).toStrictEqual([
        '3',
        'a',
        'abc',
      ]);
      expect(CrateDatasource.getIndexSuffix('abcd')).toStrictEqual([
        'ab',
        'cd',
        'abcd',
      ]);
      expect(CrateDatasource.getIndexSuffix('abcde')).toStrictEqual([
        'ab',
        'cd',
        'abcde',
      ]);
    });
  });

  describe('getReleases', () => {
    let tmpDir: DirectoryResult | null;
    let adminConfig: RepoGlobalConfig;

    beforeEach(async () => {
      tmpDir = await dir({ unsafeCleanup: true });

      adminConfig = {
        localDir: join(tmpDir.path, 'local'),
        cacheDir: join(tmpDir.path, 'cache'),
      };
      GlobalConfig.set(adminConfig);

      simpleGit.mockReset();
      memCache.init();
    });

    afterEach(async () => {
      await tmpDir?.cleanup();
      tmpDir = null;
      GlobalConfig.reset();
    });

    it('returns null for missing registry url', async () => {
      // FIXME: should not call default registry?
      httpMock.scope(baseUrl).get('/no/n_/non_existent_crate').reply(404, {});
      expect(
        await getPkgReleases({
          datasource,
          depName: 'non_existent_crate',
          registryUrls: [],
        })
      ).toBeNull();
    });

    it('returns null for invalid registry url', async () => {
      expect(
        await getPkgReleases({
          datasource,
          depName: 'non_existent_crate',
          registryUrls: ['3'],
        })
      ).toBeNull();
    });

    it('returns null for empty result', async () => {
      mockCratesApiCallFor('non_existent_crate');
      httpMock.scope(baseUrl).get('/no/n_/non_existent_crate').reply(200, {});
      expect(
        await getPkgReleases({
          datasource,
          depName: 'non_existent_crate',
          registryUrls: ['https://crates.io'],
        })
      ).toBeNull();
    });

    it('returns null for missing fields', async () => {
      mockCratesApiCallFor('non_existent_crate');
      httpMock
        .scope(baseUrl)
        .get('/no/n_/non_existent_crate')
        .reply(200, undefined);
      expect(
        await getPkgReleases({
          datasource,
          depName: 'non_existent_crate',
          registryUrls: ['https://crates.io'],
        })
      ).toBeNull();
    });

    it('returns null for empty list', async () => {
      mockCratesApiCallFor('non_existent_crate');
      httpMock.scope(baseUrl).get('/no/n_/non_existent_crate').reply(200, '\n');
      expect(
        await getPkgReleases({
          datasource,
          depName: 'non_existent_crate',
          registryUrls: ['https://crates.io'],
        })
      ).toBeNull();
    });

    it('returns null for 404', async () => {
      httpMock.scope(baseUrl).get('/so/me/some_crate').reply(404);
      expect(
        await getPkgReleases({
          datasource,
          depName: 'some_crate',
          registryUrls: ['https://crates.io'],
        })
      ).toBeNull();
    });

    it('throws for 5xx', async () => {
      httpMock.scope(baseUrl).get('/so/me/some_crate').reply(502);
      await expect(
        getPkgReleases({
          datasource,
          depName: 'some_crate',
          registryUrls: ['https://crates.io'],
        })
      ).rejects.toThrow(EXTERNAL_HOST_ERROR);
    });

    it('returns null for unknown error', async () => {
      httpMock.scope(baseUrl).get('/so/me/some_crate').replyWithError('');
      expect(
        await getPkgReleases({
          datasource,
          depName: 'some_crate',
          registryUrls: ['https://crates.io'],
        })
      ).toBeNull();
    });

    it('processes real data: libc', async () => {
      mockCratesApiCallFor('libc', Fixtures.get('libc.json'));

      httpMock
        .scope(baseUrl)
        .get('/li/bc/libc')
        .reply(200, Fixtures.get('libc'));
      const res = await getPkgReleases({
        datasource,
        depName: 'libc',
        registryUrls: ['https://crates.io'],
      });
      expect(res).toMatchSnapshot();
      expect(res).not.toBeNull();
      expect(res).toBeDefined();
    });

    it('processes real data: amethyst', async () => {
      mockCratesApiCallFor('amethyst', Fixtures.get('amethyst.json'));

      httpMock
        .scope(baseUrl)
        .get('/am/et/amethyst')
        .reply(200, Fixtures.get('amethyst'));
      const res = await getPkgReleases({
        datasource,
        depName: 'amethyst',
        registryUrls: ['https://crates.io'],
      });
      expect(res).toMatchSnapshot();
      expect(res).not.toBeNull();
      expect(res).toBeDefined();
    });

    it('refuses to clone if allowCustomCrateRegistries is not true', async () => {
      const { mockClone } = setupGitMocks();

      const url = 'https://dl.cloudsmith.io/basic/myorg/myrepo/cargo/index.git';
      const res = await getPkgReleases({
        datasource,
        depName: 'mypkg',
        registryUrls: [url],
      });
      expect(mockClone).toHaveBeenCalledTimes(0);
      expect(res).toBeNull();
    });

    it('clones cloudsmith private registry', async () => {
      const { mockClone } = setupGitMocks();
      GlobalConfig.set({ ...adminConfig, allowCustomCrateRegistries: true });
      const url = 'https://dl.cloudsmith.io/basic/myorg/myrepo/cargo/index.git';
      const res = await getPkgReleases({
        datasource,
        depName: 'mypkg',
        registryUrls: [url],
      });
      expect(mockClone).toHaveBeenCalled();
      expect(res).toMatchSnapshot();
      expect(res).not.toBeNull();
      expect(res).toBeDefined();
    });

    it('clones other private registry', async () => {
      const { mockClone } = setupGitMocks();
      GlobalConfig.set({ ...adminConfig, allowCustomCrateRegistries: true });
      const url = 'https://github.com/mcorbin/testregistry';
      const res = await getPkgReleases({
        datasource,
        depName: 'mypkg',
        registryUrls: [url],
      });
      expect(mockClone).toHaveBeenCalled();
      expect(res).toMatchSnapshot();
      expect(res).not.toBeNull();
      expect(res).toBeDefined();
    });

    it('clones once then reuses the cache', async () => {
      const { mockClone } = setupGitMocks();
      GlobalConfig.set({ ...adminConfig, allowCustomCrateRegistries: true });
      const url = 'https://github.com/mcorbin/othertestregistry';
      await getPkgReleases({
        datasource,
        depName: 'mypkg',
        registryUrls: [url],
      });
      await getPkgReleases({
        datasource,
        depName: 'mypkg',
        registryUrls: [url],
      });
      expect(mockClone).toHaveBeenCalledTimes(1);
    });

    it('guards against race conditions while cloning', async () => {
      const { mockClone } = setupGitMocks(250);
      GlobalConfig.set({ ...adminConfig, allowCustomCrateRegistries: true });
      const url = 'https://github.com/mcorbin/othertestregistry';

      await Promise.all([
        getPkgReleases({
          datasource,
          depName: 'mypkg',
          registryUrls: [url],
        }),
        getPkgReleases({
          datasource,
          depName: 'mypkg-2',
          registryUrls: [url],
        }),
      ]);

      await getPkgReleases({
        datasource,
        depName: 'mypkg-3',
        registryUrls: [url],
      });

      expect(mockClone).toHaveBeenCalledTimes(1);
    });

    it('returns null when git clone fails', async () => {
      setupErrorGitMock();
      GlobalConfig.set({ ...adminConfig, allowCustomCrateRegistries: true });
      const url = 'https://github.com/mcorbin/othertestregistry';

      const result = await getPkgReleases({
        datasource,
        depName: 'mypkg',
        registryUrls: [url],
      });
      const result2 = await getPkgReleases({
        datasource,
        depName: 'mypkg-2',
        registryUrls: [url],
      });

      expect(result).toBeNull();
      expect(result2).toBeNull();
    });
  });

  describe('fetchCrateRecordsPayload', () => {
    it('rejects if it has neither clonePath nor crates.io flavor', async () => {
      const info: RegistryInfo = {
        rawUrl: 'https://example.com',
        url: new URL('https://example.com'),
        flavor: 'cloudsmith',
      };
      const crateDatasource = new CrateDatasource();
      await expect(
        crateDatasource.fetchCrateRecordsPayload(info, 'benedict')
      ).toReject();
    });
  });
});
