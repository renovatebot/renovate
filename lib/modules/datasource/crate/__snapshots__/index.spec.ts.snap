// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/datasource/crate/index getReleases clones cloudsmith private registry 1`] = `
{
  "dependencyUrl": "https://cloudsmith.io/~myorg/repos/myrepo/packages/detail/cargo/mypkg",
  "registryUrl": "https://dl.cloudsmith.io/basic/myorg/myrepo/cargo/index.git",
  "releases": [
    {
      "version": "0.1.0",
    },
    {
      "version": "0.1.1",
    },
  ],
}
`;

exports[`modules/datasource/crate/index getReleases clones other private registry 1`] = `
{
  "dependencyUrl": "https://github.com/mcorbin/testregistry/mypkg",
  "registryUrl": "https://github.com/mcorbin/testregistry",
  "releases": [
    {
      "version": "0.1.0",
    },
    {
      "version": "0.1.1",
    },
  ],
}
`;

exports[`modules/datasource/crate/index getReleases processes real data: amethyst 1`] = `
{
  "dependencyUrl": "https://crates.io/crates/amethyst",
  "homepage": "https://amethyst.rs/",
  "registryUrl": "https://crates.io",
  "releases": [
    {
      "version": "0.1.0",
    },
    {
      "version": "0.1.1",
    },
    {
      "version": "0.1.3",
    },
    {
      "version": "0.1.4",
    },
    {
      "version": "0.2.1",
    },
    {
      "version": "0.3.0",
    },
    {
      "version": "0.3.1",
    },
    {
      "version": "0.4.0",
    },
    {
      "version": "0.4.1",
    },
    {
      "version": "0.4.2",
    },
    {
      "version": "0.4.3",
    },
    {
      "version": "0.5.0",
    },
    {
      "version": "0.5.1",
    },
    {
      "version": "0.6.0",
    },
    {
      "version": "0.7.0",
    },
    {
      "version": "0.8.0",
    },
    {
      "version": "0.9.0",
    },
    {
      "version": "0.10.0",
    },
    {
      "isDeprecated": true,
      "version": "0.10.1",
    },
  ],
  "sourceUrl": "https://github.com/amethyst/amethyst",
}
`;

exports[`modules/datasource/crate/index getReleases processes real data: libc 1`] = `
{
  "dependencyUrl": "https://crates.io/crates/libc",
  "registryUrl": "https://crates.io",
  "releases": [
    {
      "version": "0.1.0",
    },
    {
      "version": "0.1.1",
    },
    {
      "version": "0.1.2",
    },
    {
      "version": "0.1.3",
    },
    {
      "version": "0.1.4",
    },
    {
      "version": "0.1.5",
    },
    {
      "version": "0.1.6",
    },
    {
      "version": "0.1.7",
    },
    {
      "version": "0.1.8",
    },
    {
      "isDeprecated": true,
      "version": "0.1.9",
    },
    {
      "version": "0.1.10",
    },
    {
      "isDeprecated": true,
      "version": "0.1.11",
    },
    {
      "version": "0.1.12",
    },
    {
      "version": "0.2.0",
    },
    {
      "version": "0.2.1",
    },
    {
      "version": "0.2.2",
    },
    {
      "version": "0.2.3",
    },
    {
      "version": "0.2.4",
    },
    {
      "version": "0.2.5",
    },
    {
      "version": "0.2.6",
    },
    {
      "version": "0.2.7",
    },
    {
      "version": "0.2.8",
    },
    {
      "version": "0.2.9",
    },
    {
      "version": "0.2.10",
    },
    {
      "version": "0.2.11",
    },
    {
      "version": "0.2.12",
    },
    {
      "version": "0.2.13",
    },
    {
      "version": "0.2.14",
    },
    {
      "version": "0.2.15",
    },
    {
      "version": "0.2.16",
    },
    {
      "version": "0.2.17",
    },
    {
      "version": "0.2.18",
    },
    {
      "version": "0.2.19",
    },
    {
      "version": "0.2.20",
    },
    {
      "version": "0.2.21",
    },
    {
      "version": "0.2.22",
    },
    {
      "version": "0.2.23",
    },
    {
      "version": "0.2.24",
    },
    {
      "version": "0.2.25",
    },
    {
      "version": "0.2.26",
    },
    {
      "version": "0.2.27",
    },
    {
      "version": "0.2.28",
    },
    {
      "version": "0.2.29",
    },
    {
      "version": "0.2.30",
    },
    {
      "version": "0.2.31",
    },
    {
      "version": "0.2.32",
    },
    {
      "version": "0.2.33",
    },
    {
      "version": "0.2.34",
    },
    {
      "version": "0.2.35",
    },
    {
      "version": "0.2.36",
    },
    {
      "version": "0.2.37",
    },
    {
      "version": "0.2.38",
    },
    {
      "version": "0.2.39",
    },
    {
      "version": "0.2.40",
    },
    {
      "version": "0.2.41",
    },
    {
      "version": "0.2.42",
    },
    {
      "version": "0.2.43",
    },
    {
      "version": "0.2.44",
    },
    {
      "version": "0.2.45",
    },
    {
      "version": "0.2.46",
    },
    {
      "version": "0.2.47",
    },
    {
      "version": "0.2.48",
    },
    {
      "version": "0.2.49",
    },
    {
      "version": "0.2.50",
    },
    {
      "version": "0.2.51",
    },
  ],
  "sourceUrl": "https://github.com/rust-lang/libc",
}
`;
